silhouette/app/com/mohiva/play/silhouette/api/services/AuthenticatorService.scala

```diff
@@ -20,9 +20,52 @@
 package com.mohiva.play.silhouette.api.services
 
 import com.mohiva.play.silhouette.api.{ Authenticator, LoginInfo }
+import play.api.libs.concurrent.Execution.Implicits._
 import play.api.mvc.{ Headers, RequestHeader, Result }
 
 import scala.concurrent.Future
+import scala.language.implicitConversions
+import scala.util.Try
+
+/**
+ * A marker result which indicates that an operation on an authenticator was processed and
+ * therefore it shouldn't updated automatically.
+ *
+ * Due the fact that the update method gets called on every subsequent request to update the
+ * authenticator related data in the backing store and in the result, it isn't possible to
+ * discard or renew the authenticator simultaneously. This is because the "update" method would
+ * override the result created by the "renew" or "discard" method, because it will be executed
+ * as last in the chain.
+ *
+ * As example:
+ * If we discard the session in a Silhouette action then it will be removed from session. But
+ * at the end the update method will embed the session again, because it gets called with the
+ * result of the action.
+ *
+ * @param result The result to wrap.
+ */
+class AuthenticatorResult(result: Result) extends Result(result.header, result.body, result.connection)
+
+/**
+ * The companion object.
+ */
+object AuthenticatorResult {
+
+  /**
+   * Converts a [[Result]] into [[AuthenticatorResult]]
+   */
+  implicit def toAuthenticatorResult(result: Result): AuthenticatorResult = apply(result)
+  implicit def toAuthenticatorResultF(result: Future[Result]): Future[AuthenticatorResult] = result.map(apply)
+  implicit def toAuthenticatorResultT(result: Try[Result]): Try[AuthenticatorResult] = result.map(apply)
+
```

Comment:
The only other thing I can think of is to use an action wrapper to unwrap the AuthenticatorResult




the type signature for WithAuthenticatorResult.apply would be:


(Unit => Future[AuthenticatorResult]) => Future[Result]




And be used like so:


```scala


*Action.async {

  WithAuthenticatorResult {

     ....

     authenticatorService.discard(...)

  }


```






What do you think?

On Tue, Apr 7, 2015 at 7:24 AM, Christian Kaps <notifications@github.com>
wrote:

>> + * @param result The result to wrap.
>> + */
>> +class AuthenticatorResult(result: Result) extends Result(result.header, result.body, result.connection)
>> +
>> +/**
>> + * The companion object.
>> + */
>> +object AuthenticatorResult {
>> +
>> +  /**
>> +   * Converts a [[Result]] into [[AuthenticatorResult]]
>> +   */
>> +  implicit def toAuthenticatorResult(result: Result): AuthenticatorResult = apply(result)
>> +  implicit def toAuthenticatorResultF(result: Future[Result]): Future[AuthenticatorResult] = result.map(apply)
>> +  implicit def toAuthenticatorResultT(result: Try[Result]): Try[AuthenticatorResult] = result.map(apply)
>> +
> Yea, thanks! Had a nice time with my wife and the kids.
>> That test doesn't reflect the problem...addSession doesn't copy the Result like the real Result does.
> Ahh, now I see the problem.
>> Is it considered bad form to override a case class' copy method to preserve the subclass?
> I think yes.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/mohiva/play-silhouette/pull/323/files#r27869279